### PR TITLE
Allow CRI-O to manage IPC and UTS namespaces

### PR DESCRIPTION
**Pull request by haircommander**
This includes:
 - change the options `ManageNetworkNSLifecycle` to `ManageNSLifecycle`
 - add a new set of files: `internal/lib/sandbox/namespaces*` that takes care of namespace related functionality
 - create a generic NamespaceIface interface for interacting with all three kinds of namespaces
 - refactor some of runPodSandbox to reduce cyclomatic complexity

still TODO: adapt the unit tests to use IPC and UTS as well

### CODE REVIEW

**Comment by mrunalp** [comment]
We can do this for IPC and UTS but not for PID. PID requires the first pid in the namespace to be alive. 
So, once all this is in place, we should need a pod image (pause) only when pid namespace is shared. For the other cases, just the pinned namespaces are enough.

**Comment by haircommander** [comment]
Fixed!

**Comment by haircommander** [comment]
I'm doing some CI runs with `manage_ns_lifecycle = true`. We probably shouldn't merge it
/hold

File: internal/lib/sandbox/namespaces_linux.go [hunk 1]
```diff
@@ -11,68 +11,183 @@ import (
...
-	return &NetNs{netNS: netNS, closed: false, restored: true}, nil
+		var origNS nspkg.NetNS
+		origNS, err = nspkg.GetNS(getCurrentThreadNSPath(nsType))
+		if err != nil {
+			return
+		}
+		defer origNS.Close()
+
+		// create a new netns on the current thread
+		err = unix.Unshare(flag)
```
Comment (mrunalp) [inline internal/lib/sandbox/namespaces_linux.go]: We should unshare all 3 namespaces together for performance.

**Comment by haircommander** [comment]
closing in favor of https://github.com/cri-o/cri-o/pull/3048

**Pull request closed without merging**